lib/datadog/core/configuration/option.rb

```diff
@@ -9,17 +9,36 @@ class Option
         attr_reader \
           :definition
 
+        # Option setting precedence. Higher number means higher precedence.
+        module Precedence
+          REMOTE_CONFIGURATION = 2
+          PROGRAMMATIC = 1
+          DEFAULT = 0
+        end
+
         def initialize(definition, context)
           @definition = definition
           @context = context
           @value = nil
           @is_set = false
+
+          # Lowest precedence, to allow for `#set` to always succeed for a brand new `Option` instance.
+          @precedence_set = Precedence::DEFAULT
         end
 
-        def set(value)
+        # Overrides the current value for this option if the `precedence` is equal or higher than
+        # the previously set value.
+        # The first call to `#set` will always store the value regardless of precedence.
+        #
+        # @param value [Object] the new value to be associated with this option
+        # @param precedence [Precedence] from what precedence order this new value comes from
+        def set(value, precedence: Precedence::PROGRAMMATIC)
```

Comment:
Since this API is not public (...I think...?) perhaps it would make sense to not have a default for `precedence:`? I think anyway you've covered all callers already :)

Reply:
I think the default being "the user will override anything, except Remote Configuration" is the proper default here.

I don't like that this is a public API, but hiding it also prevents users writing extensions from being able to set Precedence::DEFAULT values, which is a valid use case.

I don't think this is the perfect API, but it currently models correctly the model we are going for for remote vs local settings.